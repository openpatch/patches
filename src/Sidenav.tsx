/** @jsx jsx */
import { Fragment, useState } from "react";
import { jsx, css } from "@emotion/core";
import { Box, BoxProps } from "./Box";
import { Theme } from "./themes/types";

type SectionProps = {
  label: string;
  href: string;
  active?: boolean;
  links?: LinkProps[];
};

type LinkProps = {
  href: string;
  label: string;
  active?: boolean;
};

export type SidenavProps = {
  sections: SectionProps[];
  position?: BoxProps["position"];
};

const Link = ({ href, label, active }: LinkProps) => {
  return (
    <li className={active ? "current" : undefined}>
      <a href={href}>{label}</a>
    </li>
  );
};

const Section = (props: SectionProps) => {
  const { label, links, href, active } = props;
  return (
    <li className={active ? "active" : undefined}>
      <a href={href}>{label}</a>
      <ul>
        {links?.map((l) => (
          <Link key={l.label} {...l} />
        ))}
      </ul>
    </li>
  );
};

export const Sidenav = ({ sections, position = "initial" }: SidenavProps) => {
  const [selected, setSelected] = useState(0);
  return (
    <Box
      position={position}
      css={(theme: Theme) => [
        css`
          top: 10px;
          background-color: ${theme.colors.card};
          border-radius: 6px;
          overflow: hidden;
          border: 1px solid ${theme.colors.neutral["100"]};

          ::selection,
          ::marker {
            background: ${theme.colors.primary["500"]};
            color: ${theme.colors.text};
          }
          & li {
            border-top: 1px solid ${theme.colors.neutral["100"]};
            margin-top: -1px;
            position: relative;
            z-index: 1;
          }

          & .top a {
            border-left: 2px solid transparent;
            color: ${theme.colors.neutral["700"]};
            text-decoration: none;
            display: block;
            font-weight: 500;
            padding: 8px 16px;
            transition: background-color 0.05s ease;
          }

          & .top ul a {
            font-size: 14px;
            padding-left: 24px;
          }

          & a:hover {
            background-color: rgba(14, 30, 37, 0.03);
            color: ${theme.colors.text};
            text-decoration: none;
          }

          & ul {
            background: rgba(14, 30, 37, 0.02);
            margin: 0;
            padding: 0;
            display: block;
            list-style: none;
          }

          & .active > a {
            border-left-color: ${theme.colors.primary["500"]};
            color: ${theme.colors.text};
            font-weight: 700;
          }
          & .current > a {
            color: ${theme.colors.text};
          }
        `,
      ]}
      as="nav"
      aria-label="page-menu"
    >
      <ul className="top">
        {sections.map((s) => (
          <Section key={s.label} {...s} />
        ))}
      </ul>
    </Box>
  );
};
