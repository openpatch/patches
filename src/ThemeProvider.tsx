import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
