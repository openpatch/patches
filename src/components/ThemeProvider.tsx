import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
