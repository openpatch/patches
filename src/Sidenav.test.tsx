import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "./ThemeProvider";
import { Default } from "./Sidenav.stories";

const renderWithTheme = () =>
  render(
    <ThemeProvider>
      <Default />
    </ThemeProvider>
  );

test("should render", () => {
  renderWithTheme();
});

test("should match snapshot", () => {
  const { container } = renderWithTheme();
  expect(container).toMatchSnapshot();
});
