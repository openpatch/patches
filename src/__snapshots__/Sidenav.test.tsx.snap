// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  position: initial;
  top: 10px;
  background-color: #fff;
  border-radius: 6px;
  overflow: hidden;
  border: 1px solid #E4E7EB;
}

.emotion-0::selection,
.emotion-0::marker {
  background: #27AB83;
  color: #1F2933;
}

.emotion-0 li {
  border-top: 1px solid #E4E7EB;
  margin-top: -1px;
  position: relative;
  z-index: 1;
}

.emotion-0 .top a {
  border-left: 2px solid transparent;
  color: #3E4C59;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  font-weight: 500;
  padding: 8px 16px;
  -webkit-transition: background-color 0.05s ease;
  transition: background-color 0.05s ease;
}

.emotion-0 .top ul a {
  font-size: 14px;
  padding-left: 24px;
}

.emotion-0 a:hover {
  background-color: rgba(14, 30, 37, 0.03);
  color: #1F2933;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0 ul {
  background: rgba(14, 30, 37, 0.02);
  margin: 0;
  padding: 0;
  display: block;
  list-style: none;
}

.emotion-0 .active>a {
  border-left-color: #27AB83;
  color: #1F2933;
  font-weight: 700;
}

.emotion-0 .current>a {
  color: #1F2933;
}

<div>
  <div
    id="patches-root"
  >
    <nav
      aria-label="page-menu"
      class="emotion-0 emotion-1"
    >
      <ul
        class="top"
      >
        <li
          class="active"
        >
          <a
            href="#"
          >
            Section 1
          </a>
          <ul>
            <li
              class="current"
            >
              <a
                href="#"
              >
                bal1
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                bal
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            href="#"
          >
            Section 2
          </a>
          <ul />
        </li>
        <li>
          <a
            href="#"
          >
            Link 1
          </a>
          <ul />
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
